x-pack/plugins/infra/server/lib/alerting/metric_threshold/metric_threshold_executor.ts

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { mapValues } from 'lodash';
+import { i18n } from '@kbn/i18n';
+import { InfraDatabaseSearchResponse } from '../../adapters/framework/adapter_types';
+import { createAfterKeyHandler } from '../../../utils/create_afterkey_handler';
+import { getAllCompositeData } from '../../../utils/get_all_composite_data';
+import { networkTraffic } from '../../../../common/inventory_models/shared/metrics/snapshot/network_traffic';
+import { MetricExpressionParams, Comparator, AlertStates } from './types';
+import { AlertServices, AlertExecutorOptions } from '../../../../../alerting/server';
+
+interface Aggregation {
+  aggregatedIntervals: {
+    buckets: Array<{ aggregatedValue: { value: number }; doc_count: number }>;
+  };
+}
+
+interface CompositeAggregationsResponse {
+  groupings: {
+    buckets: Aggregation[];
+  };
+}
+
+const getCurrentValueFromAggregations = (
+  aggregations: Aggregation,
+  aggType: MetricExpressionParams['aggType']
+) => {
+  try {
+    const { buckets } = aggregations.aggregatedIntervals;
+    if (!buckets.length) return null; // No Data state
+    const mostRecentBucket = buckets[buckets.length - 1];
+    if (aggType === 'count') {
+      return mostRecentBucket.doc_count;
+    }
+    const { value } = mostRecentBucket.aggregatedValue;
+    return value;
+  } catch (e) {
+    return undefined; // Error state
+  }
+};
+
+const getParsedFilterQuery: (
+  filterQuery: string | undefined
+) => Record<string, any> = filterQuery => {
+  if (!filterQuery) return {};
+  try {
+    return JSON.parse(filterQuery).bool;
+  } catch (e) {
+    return {
+      query_string: {
+        query: filterQuery,
+        analyze_wildcard: true,
+      },
+    };
+  }
+};
+
+export const getElasticsearchMetricQuery = (
```

Comment:
This function was abstracted out of `getMetric` so that we can export it for API integration tests